core-processor/src/main/java/io/micronaut/inject/ast/MemberElement.java

```diff
@@ -104,6 +94,61 @@ default boolean isAccessible() {
         return isAccessible(getOwningType());
     }
 
+    /**
+     * Returns whether this member element can be invoked or retrieved at runtime.
+     * It can be accessible by a simple invocation
+     *
+     * <p>This method uses {@link #isReflectionRequired()} with a checks if the reflection access is allowed.
+     * By checking for {@link io.micronaut.core.annotation.ReflectiveAccess} annotation.
+     * </p>
+     *
+     * @param callingType The calling type
+     * @param allowReflection If reflection invocation can be used
+     * @return Will return {@code true} if is accessible.
+     * @since 4.3.0
+     */
+    default boolean isAccessible(@NonNull ClassElement callingType, boolean allowReflection) {
+        if (isPublic()) {
+            return true;
+        }
+        if (isProtected()) {
+            // the declaring type might be a super class in which
+            // case if the super class is in a different package then
+            // the method or field is not visible and hence reflection is required
+            final ClassElement declaringType = getDeclaringType();
+            String packageName = declaringType.getPackageName();
+            if (!packageName.equals(callingType.getPackageName())) {
+                return allowReflection && hasAnnotation(ReflectiveAccess.class);
+            }
+            return true;
+        }
+        if (isPackagePrivate()) {
+            // the declaring type might be a super class in which
+            // case if the super class is in a different package then
+            // the method or field is not visible and hence reflection is required
+            final ClassElement declaringType = getDeclaringType();
+            String packageName = declaringType.getPackageName();
+            if (!packageName.equals(callingType.getPackageName())) {
+                return allowReflection && hasAnnotation(ReflectiveAccess.class);
+            }
+            if (isPackagePrivate()) {
```

Comment:
@dstepanov Do we need one more check for `isPackagePrivate()` here?

Reply:
Yeah, I tried to refactor and failed :) I will give it one more chance.